### PR TITLE
Benchmark Tests for FNO and DeepONets

### LINKED ISSUE #13: Performance Benchmarks
We need to compare with

1. Old `NeuralOperators.jl`
2. Pytorch Neural Operators https://github.com/neuraloperator/neuraloperator
3. Some Jax version?

### PR DESCRIPTION
**Pull request by ayushinav**
To fix #13 

* [x] comparison with [neuraloperators](https://github.com/neuraloperator/neuraloperator)
* [x] comparison with [deepxde](https://github.com/lululxvi/deepxde)
* [x] comparison with Flux variant
* [ ] GPU comparisons 

I had some issues with CUDA and all when installing torch CUDA toolkit.

### CODE REVIEW

**Comment by ayushinav** [comment]
I guess a more appropriate place for this would be `SciMLBenchmarks`?

**Comment by avik-pal** [comment]
Yes the CPU ones can go to SciMLBenchmarks. How are the julia native ones so slow? Did you run the profiler to see where the bottlenecks are?

**Comment by ayushinav** [comment]
For DeepONet, most of the time goes to the layers, iiuc
image


For FNOs, ig ffts are a bit expensive
image

**Comment by avik-pal** [comment]
Couple of things to check:

1. Load `MKL.jl` before calling the neural networks
2. Is the number of threads consistent between julia and python?

**Comment by avik-pal** [comment]
Also that `compile` function might be compiling the torch model with dynamo. It is pretty much impossible to beat that with Lux running in eager mode. You could try and compile the DeepONet model with https://github.com/EnzymeAD/Reactant.jl/pull/55 and see the performance.

**Comment by avik-pal** [comment]
Also looking at the plots you are on quite an old version of LuxLib, update it, and it should address some performance

**Comment by avik-pal** [comment]
@ayushinav can you install https://github.com/LuxDL/LuxLib.jl/pull/111 and let me know how the performance is?

**Comment by avik-pal** [comment]
I checked the recent lux releases. The current problems are

1. `__project` isn't fast enough. This needs to be rewritten to use LoopVectorization on CPU
2. `batched_mul` -- I am looking into this in LuxLib, this is quite easy to fix
3. allocations are expensive. Nothing much we can do there honestly.

**Comment by ChrisRackauckas** [comment]
Can we make this into a SciMLBenchmarks script? That will be easier to maintain in the long run.

We can make that support GPU

**Comment by avik-pal** [comment]
Also for python setup a `pyproject.toml` file so that setting up the benchmarks is easier

File: bench/pytorch.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+y = torch.rand(batch_size, dim_y)
+output = torch.rand(batch_size, dim_y)
+
+u_ = torch.rand(8, m)
+y_ = torch.rand(8, dim_y)
+output_ = torch.rand(8, dim_y)
+
+data = dde.data.Triple(
+    X_train= (u,y), y_train= output, X_test=(u_, y_), y_test= output_
+)
```
Comment (avik-pal) [inline bench/pytorch.py:32]: The dimensions are wrong here.

The julia version has sizes:

```julia
julia> size(u)
(32, 64)

julia> size(y)
(1, 128, 64)
```

while the pytorch versions have `(64, 32)` and `(64, 1)`

**Review by avik-pal — CHANGES_REQUESTED** [review]

**Comment by avik-pal** [comment]
Haven't looked into the FNO version much, but that will most likely need https://github.com/LuxDL/LuxLib.jl/issues/118 for performance. To summarize the issue there:

1. `gelu` and friends are surprisingly expensive operations, and fusing them into the `fused_dense` operations slows down the entire loop. But that is not true in general. For example, tanh/relu/abs, etc. can and should be fused into the main loop for performance. So we need to extend the current implementation with additional traits to see which activations can be fused for performance.
2. But the main perf gain will come from better gradient ops -- even for gelu. Currently, we fail to use LoopVectorization in that case but if we hardcode some of the common cases the performance will significantly improve.

**Comment by avik-pal** [comment]
@ayushinav can we get this finished?

**Comment by avik-pal** [comment]
@ayushinav bump

**Comment by ayushinav** [comment]
@avik-pal 
Yes, I'm on this. I need to write the gradients for the new  `__project`s using LoopVectorization. The projection becomes faster then, but there isn't a significant speed up for the overall network.

**Comment by avik-pal** [comment]
Do that in a separate PR, let's get the benchmarks aligned first. The pytorch ones use a different size it seems.

**Comment by ayushinav** [comment]
The sizes are now aligned. The python variant of DeepONet only supports 1 eval point in a batch, and the Flux variant doesn't support batching. To have the same size for inputs, I made the batch size and the eval points same to compare with both the variants.

The Flux variant of FNO only supported a fixed length of kernels, which is fixed [here](650103186fa4e69fe7a652e2e201d8a4403761f7).

The difference in size now is because python uses `(batch_size, N)` whereas julia uses `(N, batch_size)`

**Comment by avik-pal** [comment]
I am guessing the overhead in FNO is currently from fft?

**Comment by ayushinav** [comment]
Fair share of fft and matmuladd, though the share of later has increased

image

**Comment by avik-pal** [comment]
Can you also profile the backward pass for the FNO? I am surprised it is that bad

**Comment by ayushinav** [comment]
For mse loss, profiling the backward as `gradient(ps -> loss(first(model(x, ps, st)), y), ps)` 

image

**Comment by avik-pal** [comment]
using the permuted formulation it is now all just fft time in forward and backward. It is quite surprising that our FFT is so much slower than pytorch.

@ayushinav might be worth giving https://github.com/Taaitaaiger/RustFFT.jl a shot and checking the  performance on CPUs
